### PR TITLE
[PSyIR] Add support for matmul in PSyIR (issue #661)

### PR DESCRIPTION
**Pull request by rupertford**
MATMUL is currently a code block in the PSyIR. This PR adds it as a PSyIR BinaryOperation. It also renames the dynamo examples to lfric.

### CODE REVIEW

**Comment by rupertford** [comment]
Ready for review assuming travis is happy. Note the changes here are quite small, however in deciding to change the name of the dynamo directory to lfric lots of files have been moved, hence the rather large number of modified files. Note codecov seems to be in a strange state again.

**Comment by arporter** [comment]
Travis is *not* happy! Please fix...

**Comment by rupertford** [comment]
travis is now happy apart from codecov which is complaining about code I've not changed. Ready for review.

File: examples/lfric/eg15/matvec_opt.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+Optimising matvec in PSyclone is work in progress. At the moment the
+only automated optimisation possible is the kernel constant
+transformation (see eg13) and this does little to matvec as it stands.
+
+Below is a list of things that will be implemented to improve
+performance but are not yet supported as transformations in PSyclone.
+
+1) replace matmult with equivalent code
+   a) get PSyIR to recognise array slice notation
+   b) replace matmult with equivalent inline code
```
Comment (arporter) [inline examples/lfric/eg15/matvec_opt.py]: s/matmult/matmul/

File: examples/lfric/eg15/matvec_opt.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+will automate these optimisations.
+
+Optimising matvec in PSyclone is work in progress. At the moment the
+only automated optimisation possible is the kernel constant
+transformation (see eg13) and this does little to matvec as it stands.
+
+Below is a list of things that will be implemented to improve
+performance but are not yet supported as transformations in PSyclone.
+
+1) replace matmult with equivalent code
```
Comment (arporter) [inline examples/lfric/eg15/matvec_opt.py]: s/matmult/Fortran intrinsic MATMUL/

File: examples/lfric/eg15/matvec_opt.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    fortran_writer = FortranWriter()
+    for invoke in psy.invokes.invoke_list:
+        schedule = invoke.schedule
+        for kernel in schedule.coded_kernels():
+            if kernel.name.lower() == "matrix_vector_kernel_code":
+                kernel_schedule = kernel.get_kernel_schedule()
+                # optimisations will go here.
+                kernel_schedule.view()
+                result = fortran_writer(kernel_schedule)
+                print (result)
```
Comment (arporter) [inline examples/lfric/eg15/matvec_opt.py]: pylint complains about the whitespace before the bracket.

File: examples/lfric/eg15/matvec_opt.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+                # Abort after the first matrix vector kernel for the
+                # time being.
+                exit(1)
```
Comment (arporter) [inline examples/lfric/eg15/matvec_opt.py]: Please print a message (to stdout) here to explain to a casual user (or us when we've forgotten about this line) why we've suddenly aborted.

File: src/psyclone/psyir/nodes/operation.py [hunk 5]
```diff
@@ -197,7 +197,9 @@ class BinaryOperation(Operation):
         # Other Maths Operators
         'SIGN', 'MIN', 'MAX',
         # Query Operators
-        'SIZE'
+        'SIZE',
+        # Matrix and Vector Operators
```
Comment (arporter) [inline src/psyclone/psyir/nodes/operation.py]: We really need to precisely document these various intrinsics that we support - what arguments (and types) they take and what they return. However, I'm attempting to find a nice way to do that in #672 so I'll let you off here.

File: doc/user_guide/getting_going.rst [hunk 6]
```diff
@@ -338,7 +338,7 @@ There are 3 subdirectories (``dynamo``, ``gocean`` and
 PSyclone. In this case we are going to use one of the dynamo
 examples::
 
-   > cd <EGS_HOME>/examples/dynamo/eg1
```
[the remark this answers is not in the document]
Comment (rupertford) [reply doc/user_guide/getting_going.rst]: Actually there were loads of changes needed in the documentation - as well as this one. After grep'ing I think I've got them all.

File: examples/lfric/eg15/matvec_opt.py (hunk 1 above)
Comment (rupertford) [reply examples/lfric/eg15/matvec_opt.py]: whoops.

File: examples/lfric/eg15/matvec_opt.py (hunk 3 above)
Comment (rupertford) [reply examples/lfric/eg15/matvec_opt.py]: My (old?) Python2 pylint didn't complain so I missed that. Updated.

File: examples/lfric/eg15/matvec_opt.py (hunk 4 above)
Comment (rupertford) [reply examples/lfric/eg15/matvec_opt.py]: Good idea

File: src/psyclone/psyir/nodes/operation.py (hunk 5 above)
Comment (rupertford) [reply src/psyclone/psyir/nodes/operation.py]: Whoop :-)

File: examples/lfric/eg15/matvec_opt.py (hunk 2 above)
Comment (rupertford) [reply examples/lfric/eg15/matvec_opt.py]: Better thanks.

**Comment by rupertford** [comment]
I think I've addressed everything. Ready for next review.

File: doc/user_guide/examples.rst [hunk 7]
```diff
@@ -249,6 +249,24 @@ within an OpenACC Parallel directive and 4) there are no checks on
...
 i.e. support for colouring or locking is not yet implemented.
 
+Example 15: CPU Optimisation of Matvec
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Example of optimising the LFRic matvec kernel for CPUs. This is work
+in progress with the idea being that PSyclone transformations will be
+able to reproduce hand-optimised code.
+
+There is one script which, when run
```
Comment (arporter) [inline doc/user_guide/examples.rst]: This isn't formatted correctly: should be "run::"

File: examples/lfric/eg15/matvec_opt.py [hunk 8]
```diff
@@ -89,8 +89,10 @@ def trans(psy):
                 # optimisations will go here.
                 kernel_schedule.view()
                 result = fortran_writer(kernel_schedule)
-                print (result)
+                print(result)
                 # Abort after the first matrix vector kernel for the
-                # time being.
+                # time being.]
+                print ("Aborting to view the modifications to the matrix "
```
Comment (arporter) [inline examples/lfric/eg15/matvec_opt.py]: Erroneous "]" and space before brackets!

File: src/psyclone/tests/psyir/backend/fortran_test.py [hunk 9]
```diff
@@ -552,7 +552,9 @@ def test_fw_kernelschedule(fort_writer, monkeypatch):
 # within previous tests
 
 
-def test_fw_binaryoperator(fort_writer):
+@pytest.mark.parametrize("binary_intrinsic", ["mod", "matmul", "max", "min",
+                                              "sign", "sum"])
+def test_fw_binaryoperator(fort_writer, binary_intrinsic):
     '''Check the FortranWriter class binary_operation method correctly
     prints out the Fortran representation of an intrinsic. Uses sign
```
Comment (arporter) [inline src/psyclone/tests/psyir/backend/fortran_test.py]: Not any more - you've extended it to do more than that now.

File: src/psyclone/tests/psyir/backend/fortran_test.py [hunk 10]
```diff
@@ -565,14 +567,14 @@ def test_fw_binaryoperator(fort_writer):
         "subroutine tmp(a,n)\n"
         "  integer, intent(in) :: n\n"
         "  real, intent(out) :: a(n)\n"
-        "    a = sign(1.0,1.0)\n"
+        "    a = {0}(1.0,1.0)\n"
```
Comment (arporter) [inline src/psyclone/tests/psyir/backend/fortran_test.py]: This is not valid Fortran in the case of the SUM intrinsic:

    result = SUM (array [,dim] [,mask])

File: src/psyclone/tests/psyir/backend/fortran_test.py [hunk 11]
```diff
@@ -565,14 +567,14 @@ def test_fw_binaryoperator(fort_writer):
...
         "end subroutine tmp\n"
-        "end module test")
+        "end module test").format(binary_intrinsic)
     schedule = create_schedule(code, "tmp")
 
     # Generate Fortran from the PSyIR schedule
     result = fort_writer(schedule)
-    assert "a=SIGN(1.0, 1.0)" in result
+    assert "a={0}(1.0, 1.0)".format(binary_intrinsic.upper()) in result
```
Comment (arporter) [inline src/psyclone/tests/psyir/backend/fortran_test.py]: Again, this isn't valid code for the SUM intrinsic. I know we can't catch this without type information but the test itself should at least be with correct code :-)

File: doc/user_guide/examples.rst (hunk 7 above)
Comment (rupertford) [reply doc/user_guide/examples.rst]: oops

File: examples/lfric/eg15/matvec_opt.py (hunk 8 above)
Comment (rupertford) [reply examples/lfric/eg15/matvec_opt.py]: oops again.

File: src/psyclone/tests/psyir/backend/fortran_test.py (hunk 9 above)
Comment (rupertford) [reply src/psyclone/tests/psyir/backend/fortran_test.py]: Updated text

File: src/psyclone/tests/psyir/backend/fortran_test.py (hunk 10 above)
Comment (rupertford) [reply src/psyclone/tests/psyir/backend/fortran_test.py]: Created separate test.

File: src/psyclone/tests/psyir/backend/fortran_test.py (hunk 11 above)
Comment (rupertford) [reply src/psyclone/tests/psyir/backend/fortran_test.py]: Created separate test.

**Comment by rupertford** [comment]
Ready for next review.

File: examples/lfric/eg15/matvec_opt.py (hunk 8 above)
Comment (arporter) [reply examples/lfric/eg15/matvec_opt.py]: That pesky white space is still there...

File: src/psyclone/tests/psyir/backend/fortran_test.py [hunk 12]
```diff
@@ -552,10 +552,13 @@ def test_fw_kernelschedule(fort_writer, monkeypatch):
 # within previous tests
 
 
-def test_fw_binaryoperator(fort_writer):
+@pytest.mark.parametrize("binary_intrinsic", ["mod", "matmul", "max", "min",
```
Comment (arporter) [inline src/psyclone/tests/psyir/backend/fortran_test.py]: Apologies but I've just realised that `matmul` isn't valid in this test either - please could you separate it out.

File: src/psyclone/tests/psyir/backend/fortran_test.py [hunk 13]
```diff
@@ -565,14 +568,38 @@ def test_fw_binaryoperator(fort_writer):
...
         "  real, intent(out) :: a(n)\n"
-        "    a = sign(1.0,1.0)\n"
+        "    a = {0}(1.0,1.0)\n"
+        "end subroutine tmp\n"
+        "end module test").format(binary_intrinsic)
+    schedule = create_schedule(code, "tmp")
+
+    # Generate Fortran from the PSyIR schedule
+    result = fort_writer(schedule)
+    assert "a={0}(1.0, 1.0)".format(binary_intrinsic.upper()) in result
```
Comment (arporter) [inline src/psyclone/tests/psyir/backend/fortran_test.py]: We could add a compilation test here but I'm not sure it's strictly necessary (as the code should be unchanged from the input).

File: examples/lfric/eg15/matvec_opt.py (hunk 8 above)
Comment (rupertford) [reply examples/lfric/eg15/matvec_opt.py]: I forgot to change this  - I just read the first part - apologies.

File: src/psyclone/tests/psyir/backend/fortran_test.py (hunk 13 above)
Comment (rupertford) [reply src/psyclone/tests/psyir/backend/fortran_test.py]: I agree. It makes sense to test the fortran code produced by the fortran backend. I'm a bit embarrassed that this is not already done. Anyway added it for this and other operators tests.

File: src/psyclone/tests/psyir/backend/fortran_test.py (hunk 12 above)
Comment (rupertford) [reply src/psyclone/tests/psyir/backend/fortran_test.py]: I'm guessing you mean matmul(1.0, 1.0). Your suggestion of the addition of a compile test also picked this up :-)

**Comment by rupertford** [comment]
Ready for next review.

File: src/psyclone/psyir/nodes/operation.py [hunk 14]
```diff
@@ -197,7 +197,9 @@ class BinaryOperation(Operation):
         # Other Maths Operators
         'SIGN', 'MIN', 'MAX',
         # Query Operators
-        'SIZE', 'LBOUND', 'UBOUND'
+        'SIZE', 'LBOUND', 'UBOUND',
+        # Matrix and Vector Operators
+        'MATMUL'
```
Comment (arporter) [inline src/psyclone/psyir/nodes/operation.py]: Please document MATMUL in the docstring below.

Comment (rupertford) [reply src/psyclone/psyir/nodes/operation.py]: Done. Hopefully in a good enough way.

**Comment by rupertford** [comment]
Requested changed have been made. Ready for next review (once travis is happy).

**Pull request merged**